### PR TITLE
On-demand background navmesh generation

### PR DESCRIPTION
**Pull request by slipher**
Stacked on #2352.

On-demand navmesh generation in the background in sgame is ready to test out. Just load a map and add a bot. Well also you need valid navmeshes to not already exist; you can delete them from `<homepath>/game/maps/` if applicable, or edit `NAVMESHSET_VERSION` in the source to invalidate all existing meshes.

### CODE REVIEW

**Comment by illwieckz** [comment]
Impressive!

```
]/bot add * aliens 
==== Bot Navigation Initialization ==== 
[bot] Bot#1 is connecting… 
[bot] Bot#1 entered the game 
Generating navmesh for level0 
Generating bot navigation mesh for level0... 0% 
Generating navmesh for level1 
Generating navmesh for level2 
Generating navmesh for level2upg 
Generating navmesh for level3 
Generating navmesh for level3upg 
Generating navmesh for level4 
Generating navmesh for human_naked 
Generating navmesh for human_bsuit 
==== Bot Navigation Initialization ==== 
 loading navigation mesh file 'maps/rsmse-level0.navMesh'... 
 loading navigation mesh file 'maps/rsmse-level1.navMesh'... 
 loading navigation mesh file 'maps/rsmse-level2.navMesh'... 
 loading navigation mesh file 'maps/rsmse-level2upg.navMesh'... 
 loading navigation mesh file 'maps/rsmse-level3.navMesh'... 
 loading navigation mesh file 'maps/rsmse-level3upg.navMesh'... 
 loading navigation mesh file 'maps/rsmse-level4.navMesh'... 
 loading navigation mesh file 'maps/rsmse-human_naked.navMesh'... 
 loading navigation mesh file 'maps/rsmse-human_bsuit.navMesh'... 
]/bot add * humans 
[bot] Bot#2 is connecting… 
[bot] Bot#2 entered the game 
```

Do we need the `Generating bot navigation mesh for level0... 0%` log line? It seems redundant.

**Comment by illwieckz** [comment]
I have mixed feelings about “_Put helpful info in bot pings_” commit as some generic purpose or legacy tools (like [XQF](https://github.com/XQF/xqf)) detect bots by checking that player ping is zero.

**Comment by illwieckz** [comment]
For bot info, if we don't have any bot skill level above `9` we can reuse the `B` key and instead of writing a small `b` we would write the bot skill level as a digit or a letter for any other status. We may have to also edit the [Mantis](https://gitlab.com/Viech/Mantis) to keep the compatibility but no more (and the change in Mantis would be trivial).

**Comment by slipher** [comment]
> Do we need the `Generating bot navigation mesh for level0... 0%` log line? It seems redundant.

This is a message sent to all clients, while the other ones are just normal logs in the server console.

> I have mixed feelings about “_Put helpful info in bot pings_” commit as some generic purpose or legacy tools (like [XQF](https://github.com/XQF/xqf)) detect bots by checking that player ping is zero.

That wasn't working anyway as the bot pings were previously random numbers from 50 to 99. I have vague recollections that a 0 ping supposedly triggers bugs...

**Comment by illwieckz** [comment]
> > I have mixed feelings about “_Put helpful info in bot pings_” commit as some generic purpose or legacy tools (like [XQF](https://github.com/XQF/xqf)) detect bots by checking that player ping is zero.
> 
> That wasn't working anyway as the bot pings were previously random numbers from 50 to 99. I have vague recollections that a 0 ping supposedly triggers bugs...

Ah… but what do you think about `slotBots`? We already have a field to store one alphanumeric (or any other 1-byte-sized char) per bot.

**Comment by DolceTriade** [comment]
Nice will test later today!

**Comment by slipher** [comment]
> Ah… but what do you think about `slotBots`? We already have a field to store one alphanumeric (or any other 1-byte-sized char) per bot.

The ping is also a required field for every bot, and one that is already displayed to users. So to me it's nice to put something useful there without writing any new code to process it.

**Comment by illwieckz** [comment]
Once this is merged, we can do:

- https://github.com/Unvanquished/unvanquished-mapeditor-support/pull/20
- https://github.com/DaemonEngine/Urcheon/pull/59
- [tools: remove daemonmap](https://gitlab.com/xonotic/netradiant/-/merge_requests/197) xonotic/netradiant!197

File: src/sgame/sg_bot_nav.cpp
Comment (illwieckz) [inline src/sgame/sg_bot_nav.cpp:66]: At some point we may want to implement a specific ping for that, but that's not blocking this PR.

File: src/sgame/sg_bot.cpp
Comment (illwieckz) [inline src/sgame/sg_bot.cpp:411]: I prefer if you move “bot skill as bot ping” hack in another PR as I don't want to block the merging of the navmesh generation feature because of some talk on this topic.

I'm fine with “ping 998” hack though, the ping explains why the bot doesn't respond, so the reusing of that value isn't bad, and even third party software can assume the player is unavailable even without knowing what we exactly mean.

**Comment by illwieckz** [comment]
I tested it with all official maps and it works.

LGTM if `Put helpful info in bot pings` commit is moved as a separate PR so we can merge the navgen feature while we discuss about that bot ping thing in a less priory way.

**Comment by slipher** [comment]
Removed the helpful bot pings commit and add the prefix `Server: ` to the broadcast message `Server: generating bot navigation mesh for $1$... $2$%`.

**Comment by illwieckz** [comment]
One minor issue I noticed but I don't consider this blocking the merge, is that doing `navedit enable` before there is any navmesh will forbid to generate navmesh without reloading the map first:

```
]/devmap forlorn
]/navedit enable level0
==== Bot Navigation Initialization ==== 
Debug: Failed to open 'game/maps/forlorn-level0.navMesh' for reading: No such file or directory 
Warn: Cannot load navigation mesh file for level0 
Warn: Couldn't load navmeshes 

]/bot fill 3 
==== Bot Navigation Initialization ==== 
Warn: Cannot load navigation mesh file for level0 
Warn: Couldn't load navmeshes 
]/bot fill 3 
Warn: Navmesh initialization previously failed, doing nothing 
Navigation mesh files unavailable for this map 

]/devmap forlorn
]/bot fill 3 
==== Bot Navigation Initialization ==== 
Generating navmesh for level0 
[bot] Bot#1 is connecting… 
[bot] Bot#1 entered the game 
[bot] Bot#2 is connecting… 
[bot] Bot#2 entered the game 
[bot] Bot#3 is connecting… 
[bot] Bot#3 entered the game 
[bot] Bot#4 is connecting… 
[bot] Bot#4 entered the game 
[bot] Bot#5 is connecting… 
[bot] Bot#5 entered the game 
[bot] Bot#6 is connecting… 
[bot] Bot#6 entered the game 
Generating bot navigation mesh for level0... 0% 
Generating navmesh for level1 
Generating navmesh for level2 
Generating navmesh for level2upg 
Generating navmesh for level3 
Generating navmesh for level3upg 
Generating navmesh for level4 
Generating navmesh for human_naked 
Generating navmesh for human_bsuit 
```

**Review by illwieckz — APPROVED** [review]
LGTM.

**Comment by slipher** [comment]
> One minor issue I noticed but I don't consider this blocking the merge, is that doing `navedit enable` before there is any navmesh will forbid to generate navmesh without reloading the map first:

Yeah that was by design. Although I guess we could make /navedit ignore the failure init status and retry since it is in response to (presumably) a human input. The main thing is not to reattempt initialization on every `bot` command since there may be several of them in a script.

Technically you could still *generate* navmeshes with `/navgen` after a failure, but what you can't do is load them.

**Comment by necessarily-equal** [comment]
There is this warning:

```
src/sgame/sg_bot_nav.cpp:180:24: warning: enumeration value ‘GENERATING’ not handled in switch [-Wswitch]
  180 |                 switch ( G_BotSetupNav( &bot, &model->navHandle ) )
      |                        ^
```

I think it would be good to handle it, but I'm not sure how. Maybe `case navMeshStatus_t::GENERATING:  return;` to avoid requesting things twice?

**Pull request merged**
